hyperspy/tests/signal/test_fourier_transform.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2007-2016 The Hyperspy developers
+#
+# This file is part of  Hyperspy.
+#
+#  Hyperspy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+#  Hyperspy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  Hyperspy.  If not, see <http://www.gnu.org/licenses/>.
+
+import pytest
+import numpy as np
+from numpy.testing import assert_allclose
+
+from hyperspy.signals import Signal1D, Signal2D, ComplexSignal1D, ComplexSignal2D, BaseSignal
+
+
+@pytest.mark.parametrize('lazy', [True, False])
```

Comment:
Actually, as far I understood, this does not test properly lazy signals (although the object tested is an instance of `LazySignal`...), because there is no implementation of the `fft` and `ifft` methods in the `LazySignal` class. @to266 can you please confirm?

If you want to add support lazy signals, you could try to use the `map` method, which is properly implemented in the `LazySignal` class.

Reply:
I think that in the latest version of dask supported numpy functions operating of dask arrays return dask arrays. Hence, if the fft functions are supported, then it should work out of the box. If not, we should create the lazy counterparts.

Reply:
I'm not familiar with the stated above... what would be the way finding it out?

Reply:
I can confirm that it does not work out of the box. To enable lazy fft the function must simply check if ``data`` is a ``dask.array`` and, if it is, use ``dask.array.fft`` etc instead of the numpy functions.

Reply:
Does it mean that one has to use`map` with `mask.array.Array` instances? Wouldn't be better to use `map` in general i.e. for usual and lazy signals? Does `map` handles lazy signals properly?

Reply:
No, there is no need to use ``map``, it is as simply as using the fft functions from ``dask.array.fft`` instead of numpy.

Reply:
Since [dask 0.15.0](http://dask.pydata.org/en/latest/changelog.html#id3), `fft`, `fftshift` and co are implemented in dask. The dask array minimum version in setup.py will need to be updated.

Yes, normally, it should be as simple as:
```python
import dask.array as da
if isinstance(self.data, da.Array):
    fft = da.fft
else:
    fft = np.fft

# and then use `fft.fftn` or `fft.fftshift`
```

Reply:
It isn't that straight forward as that((( You will see tests failing. The point is that `dask` performs `fft` and `ifft` only on the axis that has a single chunk. Meaning one should check the chunking of the lazy signal data before fft and probably rechunk it in the case if the requirement is not met. In addition to that `fftshift` seem to rechunk the data for some reason, so ifft doesn't work out of the box that cause the test failing( I can of course rechunk it after performing `fftshift`, but not sure what would be the most elegant way of doing it and how to do it without screwing up something... So any help would be appreciated.

Reply:
CC @to266

Reply:
@to266 doesn't seem to answer, any idea how to proceed here?

Reply:
By default hyperspy guarantees that the chunk contains at least 1 signal. Therefore, hopefully the fft chunking constraint shouldn't be a problem (?) Regarding the ``fftshift`` chunking issue, are you sure that it rechunks the data? Could you provide an example of the issue using data "chunked" by hyperspy? 

By the way, as ``fftshift`` just provides a different view of the data (at least in the numpy case), what about keeping the original in ``data`` and adding a ``data_fftshifted`` attribute to ``BaseFFTSignal``? The ``data_fftshifted`` can be used for plotting and slicing. After slicing, there could be a mechanism to generate the new ``data`` from the new ``fftshifted_data``.

Reply:
Would shifting be useful for anything else than plotting? If not, we can just shift the data when we plot it -see [#1727](https://github.com/hyperspy/hyperspy/pull/1727/files#diff-64e04530fdfa59f53ebfa39bd6b4fd7cR1924).

Reply:
Yes, it's useful also for slicing and ROI selection.

Reply:
@francisco-dlp here is an example:
> In[1]:
import numpy as np
import matplotlib.pyplot as plt
import hyperspy.api as hs
import dask.array as da
In [3]:
im = hs.signals.Signal2D(np.random.random((2, 3, 4, 5)))
In [4]:
im = im.as_lazy()
In [5]:
im
Out[5]:
<LazySignal2D, title: , dimensions: (3, 2|5, 4)>
In [6]:
im.data, da.Array
isinstance(im.data, da.Array)
Out[6]:
True
In [7]:
im.data.chunks
Out[7]:
((2,), (3,), (4,), (5,))
In [8]:
axes = im.axes_manager.signal_indices_in_array
In [9]:
axes
Out[9]:
(3, 2)
In [15]:
im_fft = im._deepcopy_with_new_data(da.fft.fftshift(
                    da.fft.fftn(im.data, axes=axes), axes=axes))
In [18]:
im_fft.change_dtype(complex)
In [19]:
im_fft
Out[19]:
<LazyComplexSignal2D, title: , dimensions: (3, 2|5, 4)>
In [22]:
im_ifft = im_fft._deepcopy_with_new_data(da.fft.ifftn(da.fft.ifftshift(
                    im_fft.data, axes=axes), axes=axes))
im_ifft = im_fft._deepcopy_with_new_data(da.fft.ifftn(da.fft.ifftshift(
                    im_fft.data, axes=axes), axes=axes))
---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
<ipython-input-22-9c6b54261673> in <module>()
      1 im_ifft = im_fft._deepcopy_with_new_data(da.fft.ifftn(da.fft.fftshift(
----> 2                     im_fft.data, axes=axes), axes=axes))

> ~/anaconda/lib/python3.5/site-packages/dask/array/fft.py in func(a, s, axes)
    166         for each_axis in axes:
    167             if len(a.chunks[each_axis]) != 1:
--> 168                 raise ValueError(chunk_error % (each_axis, a.chunks[each_axis]))
    169 
    170         chunks = out_chunk_fn(a, s, axes)

> ValueError: Dask array only supports taking an FFT along an axis that 
has a single chunk. An FFT operation was tried on axis 3 
which has chunks (2, 3). To change the array's chunks use dask.Array.rechunk.

The error arises since the da.ifftshift() has rechecked the Array, see bellow:

> In [25]:
im_fft.data.chunks
Out[25]:
((2,), (3,), (4,), (5,))
In [26]:
da.fft.ifftshift(im_fft.data, axes=axes).chunks
Out[26]:
((2,), (3,), (2, 2), (3, 2))